openshift-knative-operator/pkg/monitoring/monitoring_kafka.go

```diff
@@ -0,0 +1,194 @@
+// collecting namespaced broker metrics.
+// It sets `Spec.Config["namespaced-broker-resources"] key on KnativeKafka instance, so that the KnativeOperator
+// creates the `config-namespaced-broker-resources` configmap with the given content. That content is later
+// consumed by the upstream Knative Kafka controller. It applies all the resources listed in that configmap in
+// broker's namespace, whenever a new namespaced broker is created.
+func ReconcileMonitoringForNamespacedBroker(kk *serverlessoperatorv1alpha1.KnativeKafka) error {
+
+	// For each namespaced broker dataplane, we do these:
+	// - Create a Kubernetes `Service` that makes the dataplane pods accessible by Prometheus.
+	//   That service basically integrates with the RBAC proxy sidecar in the dataplane pods
+	//
+	// - Create a `ServiceMonitor` that makes Prometheus scrape our dataplane pods.
+	//   It also sets up certificates for the RBAC proxy mentioned above.
+	//
+	// - Create a `ClusterRoleBinding` of the `rbac-proxy-reviews-prom` `ClusterRole` for the
+	//  `knative-kafka-broker-data-plane` `ServiceAccount` in the broker namespace. Otherwise, RBAC proxy cannot
+	//   authorize itself.
+	//
+	// - Set `"openshift.io/cluster-monitoring": "true",` label on the broker namespace, so that Prometheus monitors
+	//    the namespace and creates certs/secrets and also scrapes the pods
+	//
+	// While it can be outdated, here's a Gist that creates these resources manually:
+	// https://gist.github.com/aliok/1a89600db9fcec0416302148fadba5ad
+
+	additionalResources, err := createUnstructuredList(
+		serviceMonitor("receiver"),
+		serviceMonitor("dispatcher"),
+		service("receiver"),
+		service("dispatcher"),
+		clusterRoleBinding(),
+		namespace(),
+	)
+	if err != nil {
+		return err
+	}
+
+	resStr, err := commonutil.MarshalUnstructured(additionalResources)
+	if err != nil {
+		return err
+	}
+
+	if len(kk.Spec.Config) == 0 {
+		kk.Spec.Config = make(map[string]map[string]string, 1)
+	}
+
+	if len(kk.Spec.Config[namespacedBrokerResourcesConfigmapName]) == 0 {
+		kk.Spec.Config[namespacedBrokerResourcesConfigmapName] = make(map[string]string, 1)
+	}
+
+	kk.Spec.Config[namespacedBrokerResourcesConfigmapName]["resources"] = resStr
+	return nil
+}
+
+func createUnstructuredList(objs ...runtime.Object) ([]unstructured.Unstructured, error) {
+	list := make([]unstructured.Unstructured, 0)
+	for _, obj := range objs {
+		unstr, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+		if err != nil {
+			return []unstructured.Unstructured{}, err
+		}
+		list = append(list, unstructured.Unstructured{Object: unstr})
+	}
+	return list, nil
+}
+
+func namespace() *corev1.Namespace {
+	return &corev1.Namespace{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Namespace",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "{{.Namespace}}",
+			Labels: map[string]string{
+				"openshift.io/cluster-monitoring": "true",
+			},
+		},
+	}
+}
```

Comment:
The namespace is controlled by the user, are we sure we're not conflicting with whatever their config (labels/annotations, etc).

I'd honestly let them provide the `openshift.io/cluster-monitoring: true` label if they want without us assuming cluster monitoring is present (note: recently OpenShift monitoring was made an optional installation)

Reply:
good point, especially w/ the optional factor of monitoring (did not know)

Reply:
If we don't provide `openshift.io/cluster-monitoring: true`, then the monitoring operator won't create the secret, which will make the dataplane pods fail I think. It is because the RBAC proxy side car for monitoring wants that secret mounted.

Reply:
I see, I don't see logic in the existing reconciler to avoid injecting RBAC proxy when cluster monitoring is disabled

Reply:
cc @skonto

Reply:
> The namespace is controlled by the user, are we sure we're not conflicting with whatever their config (labels/annotations, etc).

BTW, this label is "applied". Other labels/annotations won't change

Reply:
Ok, then we're good

Reply:
One problem with this approach is that I cannot use user-defined project monitoring instead of cluster monitoring because the label is hardcoded, can we at least try to use whatever is set on the knative-eventing namespace instead of hardcoding it here?

https://docs.openshift.com/container-platform/4.10/monitoring/enabling-monitoring-for-user-defined-projects.html

Reply:
Alternatives:

1. We don't care about user-defined project monitoring and keep things as is
2. We provide a flag in the KnativeKafka CRD to specify the label. That would apply to all namespaces though.
3. We don't set the label on the namespace. Then we need to make sure if the label is not there, we also don't inject the RBAC proxy, create the ServiceMonitor, etc. So, user needs to label the namespace, then create the namespaced broker instance. 

Option 3) sounds reasonable in the summary above, but there are some issues:
a. This would require some work on eventing-kafka-broker, as KafkaController will need to remove the RBAC proxy sidecar when propagating the dataplane deployments if there's no monitoring label set.
b. Similarly, no monitoring label is set, then the additional resources (ServiceMonitor, ClusterRoleBinding, etc.) shouldn't be created. This is again to be checked by the KafkaController.

Any other ideas @pierDipi 

Also, I am just thinking if we can merge this PR as is and implement support for user-defined project monitoring as a follow up, although it will probably need migration code as the release cut is pretty soon.

Reply:
I'm ok with keeping as is for now, and then iterate towards 2 or 3

Reply:
OK, let's merge as is.